readthedocs/projects/tasks.py

```diff
@@ -128,30 +130,22 @@ def run(self, pk, version_pk=None, build_pk=None, record=True,
             setup_successful = self.run_setup(record=record)
             if setup_successful:
                 self.run_build(record=record, docker=docker)
-            failure = self.setup_env.failure or self.build_env.failure
         except Exception as e:  # noqa
             log.exception(
                 'An unhandled exception was raised outside the build environment',
                 extra={'tags': {'build': build_pk}}
             )
-            failure = _('Unknown error encountered. '
-                        'Please include the build id ({build_id}) in any bug reports.'.format(
-                            build_id=build_pk
-                        ))
-
-        # **Always** report build status.
-        # This can still fail if the API Is totally down, but should catch more failures
-        result = {}
-        build_updates = {'state': BUILD_STATE_FINISHED}
-        build_data = {}
-        if hasattr(self, 'build'):
-            build_data.update(self.build)
-        if failure:
-            build_updates['success'] = False
-            build_updates['error'] = failure
-        build_data.update(build_updates)
-        result = api_v2.build(build_pk).patch(build_updates)
-        return result
+            unhandled_failure = _(
+                'Unknown error encountered. '
+                'Please include the build id ({build_id}) in any bug reports.'.format(
+                    build_id=build_pk
+                ))
+        finally:
+            if unhandled_failure:
+                self.build_env.build['failure'] = unhandled_failure
+            self.build_env.update_build(BUILD_STATE_FINISHED)
+
+        return self.get_build(build_pk)
```

Comment:
It might be best to leave the return logic here, I'm not sure anything is depending on it though

Reply:
I think it's the same since the code that was here, was returning the response of the `patch` request to the build endpoint.

Reply:
Oh and i missed that we're not doing the api request here anyways, and can't get the response directly.

Instead of hitting the API again, maybe return the `build_env.build` object?

Reply:
It should be the same, yeah.